doc/aaa/TACACS+ Server Monitor.md

```diff
@@ -0,0 +1,195 @@
+# TACACS+ server monitor design
+
+## Overview
+
+SONiC device usually configured with multiple TACACS+ server, when a server is unreachable, SONiC device will try to connect with next TACACS+ server.
+
+SONiC device will communicate with TACACS+ server in following scenarios:
+1. Remote user login to SONiC device.
+2. Remote user run commands on SONiC device.
+
+There is a timeout for each server, the default value is 5 seconds, this means if the first server not reachable, SONiC device will stuck there for 5 seconds when user login or running commands.
+
+To improve this issue, SONiC will add a TACACS+ server monitor, a server unreachable or slow response will be downgrade.
+
+### Functional Requirement
+- Monit requirement:
+    - Add a new monit unit for TACACS+ server status check.
+    - Check TACACS+ server connection and write network latency data to COUNTER_DB.
+    - Write TACACS+ server unreachable message to syslog.
+    - Write TACACS+ server slow response message to syslog.
+- Hostcfgd requirement:
+    - Not change CONFIG_DB.
+    - Change server priority in config file based on TACACS+ network latency data in COUNTER_DB.
+
+### Counter DB schema
+#### TACPLUS_SERVER_LATENCY Table schema
+```
+; Key
+server_key           = String     ;  TACACS+ server’s address and port
+; Attributes
+latency              = 1*10DIGIT  ; Server network latency in MS, -1 for connect to server timeout
+```
+
+#### TACPLUS_SERVER_LATENCY Table YANG model
+```yangmodule sonic-tacplus-server-latency {
+    namespace "http://github.com/sonic-net/sonic-tacplus-server-latency";
+    prefix ssys;
+    yang-version 1.1;
+
+    revision 2023-11-01 {
+        description "Initial revision.";
+    }
+
+    container sonic-tacplus-server-latency {
+
+        container TACPLUS_SERVER_LATENCY {
+
+            list TACPLUS_SERVER_LATENCY_LIST {
+                max-elements 8;
+                key "server";
+
+                leaf server {
+                    type string;
+                    description
+                        "TACACS+ server’s address and port";
+                }
+
+                leaf latency {
+                    type uint32;
+                    description "Server network latency in MS";
+                }
+            }
+        }
+    }
+}
+```
+
+### Config DB schema
+#### TACPLUS_MONITOR Table schema
+```
+; Key
+config_key           = 'config'  ;  The configuration key
+; Attributes
+enable                   = BOOLEAN   ; Enable Monitor feature
```

Comment:
>enable [](http://example.com/codeflow?start=0&length=6)

There is a `FEATURE` table to control many optional features. And suggest we separate the TACACS monitor feature and TACACS auto downgrade feature, and control them with finer granularity.